File: 1942/nessy/point.lua
local Point = {
	__tostring = function(self) return string.format("{ x: %d y: %d }", self.x, self.y) end,

	__add = function(p1, p2)
		if type(p2) == "number" then
			return nessy.point(p1.x + p2, p1.y + p2)
		else
			return nessy.point(p1.x + p2.x, p1.y + p2.y)
		end
	end,

	__sub = function(p1, p2) 
		if type(p2) == "number" then
			return nessy.point(p1.x - p2, p1.y - p2)
		else
			return nessy.point(p1.x - p2.x, p1.y - p2.y)
		end
	end,

	__div = function(p1, p2)
		if type(p2) == "number" then
			return nessy.point(p1.x / p2, p1.y / p2)
		else
			return nessy.point(p1.x / p2.x, p1.y / p2.y)
		end
	end,

	__mul = function(p1, p2)
		if type(p2) == "number" then
			return nessy.point(p1.x * p2, p1.y * p2)
		else
			return nessy.point(p1.x * p2.x, p1.y * p2.y)
		end
	end
}

function Point:unpack()
	return self.x, self.y
end

function Point:map(func)
	return nessy.point(func(self.x), func(self.y))
end

function point(x, y)
	local point = new(Point, {
		x = x or 0,
		y = y or 0
	})

	return point
end

return point

File: nessyjs/1942/scripts/nessy/point.js
nessy.Point = function(x, y) {
	this.x = x || 0
	this.y = y || 0
	
	//console.log("new point "/* + this*/)
}

nessy.Point.prototype = {

	add: function(point) {
		return new nessy.Point(this.x + point.x, this.y + point.y)
	},
	sub: function(point) {
		this.x -= point.x
		this.y -= point.y
	},
	div: function(point) {
		return new nessy.Point(this.x / point.x, this.y / point.y)
	},
	mul: function(point) {
		this.x *= point.x
		this.y *= point.y
	},
	set: function(point) {
		this.x = point.x
		this.y = point.y
		return this
	},
	toString: function() {
		return "{ x: " + this.x + ", y: " + this.y + " }";
	}
}

nessy.Point.cast = function(obj) {
	if (obj instanceof nessy.Point)
		return obj

	return new nessy.Point(obj.x, obj.y)
}

nessy.Point.zero = new nessy.Point(0, 0)
nessy.Point.one = new nessy.Point(1, 1)